File: WorldSimulatorCPP/Human.cpp
#include "Human.h"
#include "World.h"

Human::Human(World* w, Coordinate c) {
    superStrengthTimer = 0;
    strength = 5;
    initiative = 4;
    species = "Human";
    sprite = 'H';
    age = 0;
    world = w;
    coord = c;
    dead = false;
    w->addOrganism(this);
}

Human::Human(World* w, Coordinate c, int s, int a, int sst) {
    superStrengthTimer = sst;
    strength = s;
    initiative = 4;
    species = "Human";
    sprite = 'H';
    age = a;
    world = w;
    coord = c;
    dead = false;
    w->addOrganism(this);
}

bool Human::canMove(Coordinate c) const {
    if (c.x < 0 || c.y < 0 ||
        c.x >= world->getColumns() ||
        c.y >= world->getRows() ||
        c == this->coord) // to jest potrzebne do wejœcia w pêtlê
        return false;
    else
        return true;
}

void Human::action() {
    cout << "Wcisnij W S A D lub Spacje by aktywowac zdolnosc" << endl;
	Coordinate newCoord;
    do {
        newCoord = getCoord();
        int token = 0;
        switch ((token = _getch())) {
        case W_KEY:
            newCoord.y -= 1;
            break;
        case D_KEY:
            newCoord.x += 1;
            break;
        case S_KEY:
            newCoord.y += 1;
            break;
        case A_KEY:
            newCoord.x -= 1;
            break;
        case SPACE:
            activateSuperStrength();
            break;
        }
    } while (!canMove(newCoord));
    Organism* opponent = world->findOrganismByCoordinate(newCoord);
    cout << getSpecies() << " przechodzi z pola " << getCoord() << " na pole " << newCoord << endl;
    if (opponent == nullptr)
    {
        setCoord(newCoord);
        cout << endl;
    }
    else
        collision(this, opponent);
}

void Human::activateSuperStrength() {
    if (superStrengthTimer == 0) {
        superStrengthTimer = 5;
        cout << "Aktywowano zdolnosc specjalna" << endl;
        cout << getSpecies() << " zwieksza swoja sile z " << getStrength() << " do " << getStrength() + SUPER_STRENGTH_INCREASE << endl;
        setStrength(getStrength() + SUPER_STRENGTH_INCREASE);
    }
    else {
        cout << "Jeszcze nie mozna aktywowac zdolnosci specjalnej" << endl;
    }
}

void Human::draw() const {
    cout << CYAN << sprite << RESET;
}

void Human::endTurn() {
    age += 1;
    if (superStrengthTimer != 0) // dopóki zdolnoœæ jest aktywna nale¿y zmniejszyæ timer i si³ê o 1
    {
        cout << "Sila " << getSpecies() << " spada o 1 z " << getStrength() << " do " << getStrength() - 1 << endl;
        superStrengthTimer -= 1;
        strength -= 1;
    }
}

void Human::setSuperStrengthTimer(int timer) {
    superStrengthTimer = timer;
}

int Human::getSuperStrengthTimer() const {
    return superStrengthTimer;
}

File: WorldSimulatorJAVA/Human.java
package worldsimulator;

public class Human extends Animal {
    private static final int SUPER_STRENGTH_INCREASE = 5;
	private int superStrengthTimer;
	private Directions moveDirection;
    
    public Human(World w, Coordinate c) {
    	superStrengthTimer = 0;
    	strength = Species.HUMAN.getStrength();
        initiative = Species.HUMAN.getInitiative();
        species = Species.HUMAN.getName();
        sprite = Species.HUMAN.getSprite();
        age = 0;
        world = w;
        coord = c;
        dead = false;
        world.addOrganism(this);
    }

    public Human(World w, Coordinate c, int s, int a, int ss) {
    	superStrengthTimer = ss;
    	strength = s;
    	initiative = Species.HUMAN.getInitiative();
        species = Species.HUMAN.getName();
        sprite = Species.HUMAN.getSprite();
        age = a;
        world = w;
        coord = c;
        dead = false;
        world.addOrganism(this);
    }
    
    @Override
    public void kill() {
        dead = true;
        coord = new Coordinate(-1, -1); // Setting the coordinate to (-1, -1) to signify death
        world.killHuman();
    }
    
    public boolean canHumanMove() {
    	if (moveDirection == null) {
    		return false;
    	}
    	Coordinate newCoord = new Coordinate (getCoord());
    	if (world instanceof GridWorld) {
        	switch (moveDirection) {
            case NORTH:
                newCoord.y -= 1;
                break;
            case EAST:
                newCoord.x += 1;
                break;
            case SOUTH:
                newCoord.y += 1;
                break;
            case WEST:
                newCoord.x -= 1;
                break;
			default:
				break;
            }
        } else if (world instanceof HexWorld) {
        	switch (moveDirection) {
        	case NORTH:
                newCoord.x -= 1;
                newCoord.y += 1;
                break;
            case NORTH_EAST:
            	newCoord.x -= 1;
                break;
            case NORTH_WEST:
            	newCoord.y += 1;
                break;
            case SOUTH_EAST:
            	newCoord.y -= 1;
                break;
            case SOUTH_WEST:
            	newCoord.x += 1;
                break;
            case SOUTH:
            	newCoord.x += 1;
                newCoord.y -= 1;
                break;
			default:
				break;
            }
        }
    	if (!world.checkBoundaries(newCoord) || newCoord.equals(coord)) {
            return false;
        } else {
            return true;
        }
    }
    
    @Override
    public void action() {
    	String message;
        Coordinate newCoord = new Coordinate (getCoord());
        if (world instanceof GridWorld) {
        	switch (moveDirection) {
            case NORTH:
                newCoord.y -= 1;
                break;
            case EAST:
                newCoord.x += 1;
                break;
            case SOUTH:
                newCoord.y += 1;
                break;
            case WEST:
                newCoord.x -= 1;
                break;
			default:
				break;
            }
        } else if (world instanceof HexWorld) {
        	switch (moveDirection) {
        	case NORTH:
                newCoord.x -= 1;
                newCoord.y += 1;
                break;
            case NORTH_EAST:
            	newCoord.x -= 1;
                break;
            case NORTH_WEST:
            	newCoord.y += 1;
                break;
            case SOUTH_EAST:
            	newCoord.y -= 1;
                break;
            case SOUTH_WEST:
            	newCoord.x += 1;
                break;
            case SOUTH:
            	newCoord.x += 1;
                newCoord.y -= 1;
                break;
			default:
				break;
            }
        }
        Organism opponent = world.findOrganismByCoordinate(newCoord);
        message = String.format(getSpecies() + " przechodzi z pola " + getCoord() + " na pole " + newCoord);
        world.prompt(message);
        if (opponent == null) {
            setCoord(newCoord);
            message = String.format("\n");
            world.prompt(message);
        } else {
            collision(this, opponent);
        }
    }

    public void activateSuperStrength() {
    	String message;
        if (superStrengthTimer == 0) {
            superStrengthTimer = 10;
            message = String.format("Aktywowano zdolnosc specjalna");
            world.prompt(message);
            message = String.format(getSpecies() + " zwieksza swoja sile z " + getStrength() + " do " + (getStrength() + SUPER_STRENGTH_INCREASE));
            world.prompt(message);
            setStrength(getStrength() + SUPER_STRENGTH_INCREASE);
        } else {
        	message = String.format("Jeszcze nie mozna aktywowac zdolnosci specjalnej");
        	world.prompt(message);
        }
    }

    @Override
    public void endTurn() {
        super.endTurn();
        if (superStrengthTimer > 5) {
        	String message = String.format("Sila " + getSpecies() + " spada o 1 z " + getStrength() + " do " + (getStrength() - 1));
            world.prompt(message);
            strength -= 1;
        }
        if (superStrengthTimer != 0) {
        	superStrengthTimer -= 1;
        }
        moveDirection = null;
    }

    public void setSuperStrengthTimer(int timer) {
        superStrengthTimer = timer;
    }
    
    public void setMoveDirection(Directions dir) {
        moveDirection = dir;
    }

    public int getSuperStrengthTimer() {
        return superStrengthTimer;
    }
    
    public Directions getMoveDirection() {
        return moveDirection;
    }

}
